File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/G.Стек-MaxEffective/solution.cpp
#include <iostream>
#include <vector>
#include <string>

class Node {
public:
    int value;
    Node* next = nullptr;
    Node* prev = nullptr;
    Node (int value = 0, Node* next = nullptr, Node* prev = nullptr) {
        this->value = value;
        this->next = next;
        this->prev = prev;
    };
};

class Stack {
private:
    Node* head;
    Node* tail;
    Node* maxHead;
    Node* maxTail;
    int size = 0;
public:
    Stack () {
        head = new Node();
        tail = head;
        maxHead = new Node();
        maxTail = maxHead;
    };
    void push(int x) {
        ++size;
        if (size == 1) {
            head->next = new Node(x);
            tail = head->next;
            tail->prev = head;
            maxHead->next = new Node(x);
            maxTail = maxHead->next;
            maxTail->prev = maxHead;
            return;
        }
        tail->next = new Node(x);
        tail->next->prev = tail;
        tail = tail->next;
        maxTail->next = new Node(std::max(x, maxTail->value));
        maxTail->next->prev = maxTail;
        maxTail = maxTail->next;
    }
    void pop() {
        if (!size) {
            std::cout << "error\n";
            return;
        }
        --size;
        tail = tail->prev;
        maxTail = maxTail->prev;
    }
    void get_max() {
        if (!size) {
            std::cout << "None\n";
            return;
        }
        std::cout << maxTail->value << "\n";
    }
    void top() {
        if (!size) {
            std::cout << "error\n";
            return;
        }
        std::cout << tail->value << "\n";
    }
};

void solution(int t) {
    /*
    Time Complexity O(1)
    Memory Complexity O(N)
    */
    Stack stack = Stack();
    for (int i = 0; i < t; ++i) {
        std::string method;
        std::cin >> method;
        if (method == "get_max") {
            stack.get_max();
        } else if (method == "top") {
            stack.top();
        } else if (method == "pop") {
            stack.pop();
        } else {
            int num;
            std::cin >> num;
            stack.push(num);
        }
    }
}

int main() {
    int t;
    std::cin >> t;
    solution(t);
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/2_sprint/G.Стек-MaxEffective/solution.py
import sys


class Double_node:
    def __init__(
        self, value: int = None,
        next_: 'Double_node' = None,
        prev_: 'Double_node' = None,
    ) -> None:
        self.value: int = value
        self.next_: Double_node = next_
        self.prev_: Double_node = prev_


class Stack:
    def __init__(self) -> None:
        self.head: Double_node = Double_node()
        self.tail: Double_node = self.head
        self.max_head: Double_node = Double_node()
        self.max_tail: Double_node = Double_node()
        self.size: int = 0
    def push(self, x: int) -> None:
        self.size += 1
        if self.size == 1:
            self.head.next = Double_node(x)
            self.tail = self.head.next
            self.tail.prev = self.head
            self.max_head.next = Double_node(x)
            self.max_tail = self.max_head.next
            self.max_tail.prev = self.max_head
            return
        self.tail.next = Double_node(x)
        self.tail.next.prev = self.tail
        self.tail = self.tail.next
        self.max_tail.next = Double_node(max(self.max_tail.value, x))
        self.max_tail.next.prev = self.max_tail
        self.max_tail = self.max_tail.next
    def pop(self) -> None:
        if not self.size:
            sys.stdout.write('error\n')
            return
        self.tail = self.tail.prev
        self.max_tail = self.max_tail.prev
        self.size -= 1
    def get_max(self) -> None:
        if not self.size:
            sys.stdout.write('None\n')
            return
        sys.stdout.write(str(self.max_tail.value) + '\n')
    def top(self) -> None:
        if not self.size:
            sys.stdout.write('error\n')
            return
        sys.stdout.write(str(self.tail.value) + '\n')


def solution(t: int) -> None:
    """
    Time Complexity O(1)
    Memory Complexity O(N)
    """
    stack: Stack = Stack()
    for _ in range(t):
        method: str = sys.stdin.readline().rstrip()
        if method.startswith('get_max'):
            stack.get_max()
        elif method.startswith('pop'):
            stack.pop()
        elif method.startswith('top'):
            stack.top()
        else:
            num: int = int(method.split()[-1])
            stack.push(num)

if __name__ == '__main__':
    t: int = int(sys.stdin.readline().rstrip())
    solution(t)